### PR TITLE
Workflow Form

### PR DESCRIPTION
**Pull request by aaronschachter**
- Removes debugging call that seemed to break things when on
- Adds lots of help text and form validation
- Gives `communication team` role access to add and edit SMS Workflows

### CODE REVIEW

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc [hunk 1]
```diff
@@ -9,8 +9,13 @@
  */
 function dosomething_sms_workflow_overview_page() {
   $admin_path = DOSOMETHING_SMS_WORKFLOW_ADMIN_PATH;
+
+  $help = "A SMS Workflow enables Reportbacks via MobileCommons for a Campaign.";
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:13]: Being picky but should all this markup be in a template?

Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:13]: Hmm I'd say no to this one - this is just an admin form that never is user facing.

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc [hunk 2]
```diff
@@ -46,10 +51,12 @@ function dosomething_sms_workflow_overview_page() {
...
   ));
-  $output .= '<h2>Development</h2>';
-  $output .= '<p>These should be turned off in production.</p>';
-  $config_form = drupal_get_form('dosomething_sms_workflow_config_form');
-  $output .= render($config_form);
+
+  if (user_access('administer modules')) {
+    $config_form = drupal_get_form('dosomething_sms_workflow_config_form');
+    $output .= render($config_form);
+  }
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:58]: A message when the user doesn't have the rights to access the form? I'm guessing members of the communication team will access this page who might not have the right admin rights if access ever gets more granular?

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc [hunk 3]
```diff
@@ -63,19 +70,14 @@ function dosomething_sms_workflow_form($form, &$form_state, $workflow = NULL) {
       '#value' => $workflow->id,
     );
   }
+  $desc = "The unique path for this workflow, appended to the https://www.dosomething.org/dosomething_sms URL that we use for mDatas.<p>e.g. If the endpoint is set to <em><strong>comebackclothes-reportback</strong></em>, the URL for its associated mData needs to be set to <em>https://dosomething.org/dosomething_sms/comebackclothes-reportback</em>.</p>Only use alpha-numeric characters and dashes (no spaces).";
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:73]: Template and / or external l()?

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc (hunk 2 above)
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:58]: This form is only for setting those Debug variables on though, which is kind of out of scope of what would be useful for the Communication Team.  The communication team is only able to add and edit Workflows as per this PR.

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc (hunk 3 above)
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:73]: This shouldn't be a clickable link though, this is help text to specify what the URL to set the mData is when you're setting up everything in Mobile Commons.

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc [hunk 4]
```diff
@@ -91,6 +93,14 @@ function dosomething_sms_workflow_form($form, &$form_state, $workflow = NULL) {
     // Default to 1 for new workflows.
     '#default_value' => isset($workflow->status) ? $workflow->status : 1,
   );
+  $desc = t("This title is used internally to describe this workflow. e.g. <em>Comeback Clothes 2014.</em>");
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:96]: This value is also referenced in the mData web service url, ex:
http://www.dosomething.org/dosomething_sms/imaboss-reportback
where "imaboss-reportback" would be this value.

Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:96]: You're referring to the `endpoint` form value.. which is this help text that is meant to explain that: https://github.com/DoSomething/dosomething/pull/2916#discussion_r15643135

Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:96]: This description is for the `title` field, which seems to only be used when the Workflow is constructed in `workflows.inc`

Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:96]: Yep, you're right - endpoint vs "human readable".

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc (hunk 3 above)
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:73]: Reading through the whole value for $desc this looks good.

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc [hunk 5]
```diff
@@ -138,13 +159,20 @@ function dosomething_sms_workflow_form_submit($form, &$form_state) {
...
+    '#title' => t('Development'),
+    '#collapsible' => TRUE,
+    '#collapsed' => TRUE,
+    '#description' => t('These settings should always be disabled in production.'),
+  );
+  $form['dev']['dosomething_sms_workflow_log'] =  array(
     '#type' => 'checkbox',
     '#title' => t('Enable logging'),
     '#description' => t("Logs requests to any SMS Workflow endpoint."),
     '#default_value' => variable_get('dosomething_sms_workflow_log'),
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.admin.inc:173]: Sweet!

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.features.user_permission.inc [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+/**
+ * Implements hook_user_default_permissions().
+ */
+function dosomething_sms_user_default_permissions() {
+  $permissions = array();
+
+  // Exported permission: 'administer dosomething_sms'.
+  $permissions['administer dosomething_sms'] = array(
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.features.user_permission.inc:14]: An additional admin type permission to grant to the communication team?

Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.features.user_permission.inc:14]: I'm pretty sure we don't use `administer dosomething_sms` anywhere, so i've just used this existing permission.  I've logged in as a communication team to test, and all they see is the SMS workflows page (in addition to the Opt-In config form which they already have access to)

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.module [hunk 7]
```diff
@@ -24,7 +24,7 @@ function dosomething_sms_menu() {
     'description' => 'Overview of all SMS Workflows',
     'page callback' => 'dosomething_sms_workflow_overview_page',
     'access callback' => 'user_access',
-    'access arguments' => array('administer modules'),
+    'access arguments' => array('administer dosomething_sms'),
```
Comment (deezone) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.module:27]: Ah, here are the extra rights!

**Comment by deezone** [comment]
:+1:

File: lib/modules/dosomething/dosomething_sms/dosomething_sms.info [hunk 8]
```diff
@@ -1,9 +1,9 @@
-name = DoSomething SMS
+name = dosomething_sms
```
Comment (angaither) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.info:1]: Does this module name need to change in `dosomething.info` as well?

Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_sms/dosomething_sms.info:1]: Ahh sneaky Features Export, changing shit. This is just that user facing name that appears in the Features list.  The dosomething.info uses the module machine name `dosomething_sms`.  I can change this later to make it look more pertty.

**Pull request merged**